rapticoressvc/nvd_data_helper.py

```diff
@@ -0,0 +1,223 @@
+    modification_timestamps = {}
+    try:
+        modification_timestamps = download_data_from_s3(s3_client, bucket_name, timestamps_file_key) or {}
+    except Exception as e:
+        logging.exception(e)
+    return modification_timestamps
+
+
+def upload_nvd_record(s3_client, bucket_name, cve_key, nvd_data):
+    upload_status = False
+    try:
+        upload_status = upload_data_to_s3(s3_client, bucket_name, cve_key, nvd_data)
+    except Exception as e:
+        logging.exception(e)
+    return upload_status
+
+
+def download_nvd_record(s3_client, bucket_name, cve_key):
+    nvd_record = None
+    try:
+        nvd_record = download_data_from_s3(s3_client, bucket_name, cve_key)
+    except Exception as e:
+        logging.exception(e)
+    return nvd_record
+
+
+def download_extract_zip(url):
+    response = requests.get(url)
+    with zipfile.ZipFile(io.BytesIO(response.content)) as zip_file:
+        for zip_info in zip_file.infolist():
+            with zip_file.open(zip_info) as the_file:
+                file = json.loads(the_file.read())
+                return file
+
+
+def get_nvd_file(year):
+    zip_url = f'https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-{year}.json.zip'
+    logging.debug(f"Downloading artifact {zip_url}")
+    nvd_data = download_extract_zip(zip_url)
+    return nvd_data
+
+
+def preprocess_nvd_data_for_upload(nvd_file):
+    cve_nvd_data_map = {}
+    for data in nvd_file["CVE_Items"]:
+        cve_id = data["cve"]["CVE_data_meta"]["ID"]
+        last_modified_date = data["lastModifiedDate"]
+        impact = data["impact"]
+        vector_strings = nested_lookup("vectorString", impact)
+        cve_vector = vector_strings and vector_strings[0] or None
+        base_scores = nested_lookup("baseScore", impact)
+        cve_score = base_scores and base_scores[0] or None
+        cve_nvd_data_map[cve_id] = {
+            'last_modified_date': last_modified_date,
+            'cve_vector': cve_vector,
+            'cve_score': cve_score,
+            'nvd_data': json.dumps(data),
+        }
+    return cve_nvd_data_map
+
+
+def upload_nvd_records_in_parallel(cve_nvd_data, args):
+    upload_status = False
+    try:
+        s3_client = args.get("s3_client")
+        bucket_name = args.get("bucket_name")
+        cve = list(cve_nvd_data.keys())[0]
+        data = cve_nvd_data[cve]
+
+        cve_key = f"{CVE_NVD_DATA_DIRECTORY}/{cve}"
+        upload_status = upload_nvd_record(s3_client, bucket_name, cve_key, data)
+    except Exception as e:
+        logging.exception(e)
+    return upload_status
+
+
+def upload_nvd_data_to_s3(s3_client, bucket_name, cve_nvd_data_map, modification_timestamps, max_workers=32):
+    modified_cve_list = []
+    for cve, data in cve_nvd_data_map.items():
+        try:
+            last_modified_date = data.pop("last_modified_date", "")
+            if last_modified_date and last_modified_date == modification_timestamps.get(cve, ""):
+                continue
```

Comment:
shouldn't it check if last_modified_date > modification_timestamps.get(cve, "") and then process upload